### DISCUSSION TITLE
Hello & NumberOfTremulants maximum value increase

### DISCUSSION THREAD

**Discussion by hnb2907** [GrandOrgue/grandorgue#1756] [Help]
Hi,

I've been developing a hybrid of the Barton 3-12 sample set on GO and have been having great results.  Thankyou to everyone who develops GO, it's brilliant!  I also occasionally use the Friesach and St Mary Downe organs too.
My system is a basically a modified Rodgers C445, with a PC running Ubuntu Studio.
When I've had some building work completed to the room, and tidied up, I'll post up some better details and pictures of my setup :)

So, for my Barton hybrid, I need to add some more tremulants, but it appears GO is limited to NumberOfTremulants<=10.  At the moment, I have 10 tremulants defined; 9 are "Wave" and 1 is "Synth".  

I had a quick look around the GO source code, but my programming skills aren't that good to modify it myself, or know if there are some limitations that have defined the maximum of 10?

Is it possible to increase this maximum to 32, or maybe a higher value please?

I have got a few more theatre/theater/cinema organ specific bug fixes and suggestions for improvement to GO.  I will add to the discussion another day :)

Best wishes,
ChrisC.

**Comment by larspalo** [comment]
@hnb2907 Welcome here! Increasing the number of allowed tremulants from the current 10 (documented in the help) is rather trivial, but the real question is why you need so many? The reason I ask is so that it's not a misunderstanding of how the tremulants of GO are modeled that's the root for this request.

I'm sure you know that in pipe organs the tremulants normally affect the wind supply to the pipes not by an individual rank but to a whole windchest at once (or even affecting multiple windchests!). Thus, normally the number of tremulants won't be that large. Even looking at the specification of https://gstos.org/wp-content/uploads/2017/01/Trenton_War_Memorial_3_16_Moller_Theatre_Organ-1.pdf I can only count to 7 tremulants.

But yes, I think that the number of allowed tremulants could be raised as it's very possible that someone could want to have both a synth and a wave version of every tremulant to model in the organ. In such cases where the number of tremulants in an organ is greater than five, this would currently not be possible.

**Reply by hnb2907** [reply]
Hi @larspalo 

Thanks for your quick reply :-)

Yes you are correct, and the mapping of a theatre/cinema organ to enclosures/windchests/trems is as you describe for WindchestGroup001 to WindchestGroup006.  However 007 to 010 are not exactly the standard GO use case that you might expect for a Church or Classical organ.  
Here is a quick summary, from my Redford Barton hybrid.  This is based and nearly unchanged on the brilliant work of Graham Goode and everyone involved:

- WindchestGroup001 -> Tremulant001=001 -> "Main" is 4 ranks (Flute, Flute Celeste, Viole, Viole Celeste)
- WindchestGroup002 -> Tremulant001=002 -> "Solo" is 3 ranks (Kinura, Oboe Horn, Solo String)
- WindchestGroup003 -> Tremulant001=003 -> is 2 ranks (Tuba, Diapason)
- WindchestGroup004 -> Tremulant001=004 -> is 2 ranks (Tibia Clausa 1, Tibia Clausa 2)
- WindchestGroup005 -> Tremulant001=005 -> is 1 rank (Vox Humana)
- WindchestGroup006 -> Tremulant001=006 -> is 1 rank (Clarinet)
- WindchestGroup007 -> Tremulant001=007 -> is "Traps Re-it"
- WindchestGroup008 -> Tremulant001=008 -> is "Percussion Re-it"
- WindchestGroup009 -> Tremulant001=009 -> is "Vibraphone Motor"
- WindchestGroup010 -> Tremulant001=010 -> is "Piano Sustain Pedal"

So WinchestGroup001 to WinchestGroup006 are a fairly standard layout for a pipe ranks on theatre organ of this size, of approximately 10 or 14 ranks.


Next, WinchestGroup007 to WinchestGroup010 are a GO equivalent to some special functions on a theatre organ, but are not pipe windchest tremulants as you might expect.
Here is a quick description of each...  Apologies if you already know this, I have backwards-engineered it from learning GO and how the Barton was created :)

- WindchestGroup007 controls "single tap" or "re-iterate" mode of the "traps" aka "untuned percussions" (for example tambourine, castanets, woodblock).  They are normally on the Accomp (bottom) keyboard.  There is usually a single stop tab switch that controls if all of the traps are single-tap or re-iterate, when the relevant instrument stop is enabled and a note key is held.  In this special case, the tremulant is used to select between 2 sets of sample .wav's for tremulant=off/on.  For off, a set of .wav percussive samples are used for single-tap.  For on, a 2nd set of .wav looped/non-percussive samples are used for re-iteration.  It is not normally possible in GO to mix percussive and non-percussive samples in a rank, but this method allows it.  It works perfectly in GO, and it simulates exactly how a theatre organ behaves.

- WindchestGroup008 is similar to 007.  But instead, this is for tuned percussions, usually on the Great/Solo keyboards; these can be single-tap or re-iterated when a key is held (e.g. xylophone, glockenspiel, marimba).  

- WindchestGroup009 is similar to 008.  So a theatre organ Vibraphone with the motor running the resonator butterfly valves, is called "Vibraphone"; with the motor stopped and resonator valves open, it is called "Crysoglott".  In my Barton hybrid, it is the only "Syth" tremulant.  _This was originally 2 sets of "Wave" samples, but I changed it to 1 set of samples and "Syth".  This is because I find the Barton Chrysoglott .wav samples sound better with the Synth trem applied for Vibraphone mode, compared to the real originally supplied Vibraphone with motor running .wav samples.  The effect of the vibraphone motor running for a real instrument, is an amplitude modulation tremulant, so it is a good fit for the "Synth" trem with a good choice of speed+depth settings._

- WindchestGroup010 is similar to 007.  It is used to switch between 2x .wav sample sets, for Piano without and with the sustain pedal pressed.


I'd like to add at least the following:
- WindchestGroup011 -> Tremulant001=011 -> is 1 rank (English Horn)
- WindchestGroup012 -> Tremulant001=012 -> is 1 rank (Brass and/or other Trumpet(s))
- WindchestGroup013 -> Tremulant001=013 -> is "Melotone Vibrato"
- WindchestGroup014 -> Tremulant001=014 -> is "Melotone Sustain"
...and maybe some more at a later date.

I suggested increasing to 32, because this seems an obvious power-of-2 number, but it's up for discussion.  Even on a large (say >50 rank) theatre organ, I would consider that 32 tremulants is more than enough.  I was also thinking that too many "Synth" tremulants could possibly overload the GO internal DSP to handle; hence my original comment about unknown to me technical limitations :)

For interest, if I was modelling the complete specification for the Compton at  [http://pipesinthepeaks.co.uk/](url) this would need at least 17 tremulants in GO (including the pipe windchest/ranks, traps, tuned percussions, piano, melotone etc).

Hopefully this makes some more sense of the use case for >10 tremulants :)

Cheers,
ChrisC.

**Reply by larspalo** [reply]
With #1757 we've just increased maximum number of tremulants to 999.

**Comment by hnb2907** [comment]
Excellent thank you 👍

There's no rush, please roll it in to the next normal release.

Thanks again,
Chris

**Reply by larspalo** [reply]
> There's no rush, please roll it in to the next normal release.

No other intention, but intermediate builds are already available for some time.

However, just for your information, the (mis)usage of the tremulant feature you suggest for some of the windchests to differ between percussive or looped samples could easily be acomplished by other means. A network of switches could allow to select which set of pipes (ranks that ultimately would be stops) that would be active for the selecting drawstop + another button to choose looped or non-looped version to play. We usually create a similar switch network to allow the selection of a synth tremulant version versus a wave tremulant at will.

**Reply by hnb2907** [reply]
Hi @larspalo,

Thanks for your reply.  Your comment about network of switches provided me with the clues, to implement another feature that I plan to add to the Barton hybrid.  

This new rank will have multiple stops for various 8/4/2' pitches.  It will also have another stop that selects 1 of 2 sample sets (each with its own [Rank_nnn_]).  Both sample sets have tremmed and untremmed .wav's.  So 4 sets of samples in total, all are looped/non-percussive.  I will use a new tremulant stop to select trem on/off.
I have prototyped the required [Stop_nnn_]'s with function=AND and function=NOT, and stitched together with some [Switch_nnn_]'s, to select [Rank001] or [Rank002].  Obviously the tremulant works normally in "Wave" mode on the rank selected to play.
This is working perfectly.  Thankyou for the clue!  :)  

To be fair, the (mis)use or re-use of the tremulant for the features in my previous post (implemented by the original Barton development team), is actually a easier to implement and understand.  It works exactly as needed for that case switching between 2 sample sets, so I can understand how they decided to use this method.  :)


Cheers,
Chris
